feat(chunkserver): Batch CS read jobs

dmga44 opened the pull request:
This PR targets these two features:
- make a single read job in a chunk able to read several blocks.
- make the read of the data in a single pread call for non-zoned drives.

The final goal is improving the read performance from the chunkserver side.
As a side change, a new option is added to the chunkserver list of options and a test is updated.

The idea is to squash all the commits into one.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
feat(chunkserver): Batch CS read jobs

This commit makes possible for a single hdd read job to accept, process
and retrieve the data of several contigous blocks. The packet to be
sent should contain in its OutputBuffer after a successful operation
the exact expected message for the client.

To configure this behavior the option MAX_BLOCKS_PER_HDD_READ_JOB was
added.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
feat(chunkserver): Batch CS pread calls

The goal of this change is to combine several pread calls from
contiguos blocks in a read job into a single one.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
refactor: Improve hddRead implementation

Changes:
- define startingOffsetOfBlock to make more readable some parameters.
- add isDataPreviouslyRead parameter to hddReadCrcAndBlock to avoid
duplicated code.
- make bytesRead of type uint32_t to avoid unnecessary casts.
- log case when a full block of zeroes is not being able to be put into
the buffer.
- log case when offset is not aligned to block size.
- handle the scenario when the size is not aligned to block size and
offset is aligned to block size.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
refactor: Encapsulate read packet initialization

This commit mainly creates the function prepareReadDataPacket to
encapsulate the logic of initializating a read data packet. The main
goal is to make the code more readable and maintainable.

Signed-off-by: Dave <dave@leil.io>

lgsilva3087 commented on src/protocol/cstocl.h:
```diff
@@ -31,7 +31,9 @@ namespace readData {
 	// This prefix requires CRC (uint32_t) and data (readSize * uint8_t) to be appended
-	uint32_t extraSpace = serializedSize(uint32_t()) + readSize;
+	uint32_t extraSpace = (((readSize + SFSBLOCKSIZE - 1) >> SFSBLOCKBITS) *
+	                       serializedSize(uint32_t())) +
```
consider:
sizeof(uint32_t)

Which is a constexpr always evaluated at compile time.

dmga44 replied:
Applied.

lgsilva3087 commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -100,6 +100,9 @@
 constexpr int kErrorLimit = 2;
 constexpr int kLastErrorTime = 60;
 constexpr size_t kIgnoreHeaderSize = 1;
+constexpr uint32_t startingOffsetOfBlock(uint32_t blocknum) {
+	return blocknum * SFSBLOCKSIZE;
+}
```
Please apply the code style here, the line fits in the define number of columns.

dmga44 replied:
Applied.

lgsilva3087 commented on src/chunkserver/chunkserver_entry.h:
```diff
@@ -157,6 +157,7 @@ struct ChunkserverEntry {
 	std::set<uint32_t> partiallyCompletedWrites;
 
 	/* read */
+	uint16_t maxBlocksPerHddReadJob;
```
Consider:
/// Number of contiguous blocks to read from the device in one read job.

dmga44 replied:
Applied.

lgsilva3087 reviewed (approved):
LGTM

dmga44 commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -737,23 +738,68 @@ int hddRead(uint64_t chunkId, uint32_t version, ChunkPartType chunkType,
+				uint32_t bytesRead = outputBuffer->copyIntoBuffer(OutputBuffer::BufferType::Block,
+				                                                  chunk, bytesToBeRead, offset);
+
+				if (bytesRead != bytesToBeRead) {
+					hddAddErrorAndPreserveErrno(chunk);
+					safs::log_warn("hddRead: file:{} - read error from block {} to {}",
+					               chunk->fullDataFilename().c_str(), initialBlock,
+					               initialBlock + numBlocks - 1);
+					hddReportDamagedChunk(chunk->id(), chunk->type());
+					return SAUNAFS_ERROR_IO;
```
`hddReportDamagedChunk` is actually being called.

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -404,6 +391,37 @@ void ChunkserverEntry::readFinishedCallback(uint8_t status, void *entry) {
+	auto packet = createDetachedPacketWithOutputBuffer(readDataPrefix, numBlocks);
+
+	for (uint32_t i = 0; i < numBlocks; i++) {
+		if (i > 0) {  // first block is already serialized
+			readDataPrefix.clear();
+			messageSerializer->serializePrefixOfCstoclReadData(
+			    readDataPrefix, chunkId, offset - jobOffset + (i * SFSBLOCKSIZE), SFSBLOCKSIZE);
+		}
```
The `readDataPrefix` needs to be clear before every serialization, that's what I'm ensuring here.

The calculations should not overflow because the initialization of those values make it so.

The inner csentry structures won't change during this call because only one thread access it.

dmga44 committed:
test: Test new option

This commit introduces changes in the test
test_advanced_cs_read_parameters to test the new added option.

Signed-off-by: Dave <dave@leil.io>

lgsilva3087 reviewed (approved):
Great. LGTM.

rolysr commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -136,28 +136,15 @@ MessageSerializer *MessageSerializer::getSerializer(PacketHeader::Type type) {
 
-std::unique_ptr<PacketStruct>
-ChunkserverEntry::createDetachedPacketWithOutputBuffer(
+std::unique_ptr<PacketStruct> ChunkserverEntry::createDetachedPacketWithOutputBuffer(
```
Is it needed to keep these indentation changes?

dmga44 replied:
Well, I wanted to change the signature of the method and could apply the new width rule that we have, so I did it. I think it looks better this way.

rolysr commented on src/chunkserver/network_main_thread.cc:
```diff
@@ -120,6 +120,8 @@ void mainNetworkThreadReload(void) {
 	}
 	chunkReplicatorReload();
 
+	gMaxBlocksPerHddReadJob = cfg_get_minvalue<uint16_t>(
+	    "MAX_BLOCKS_PER_HDD_READ_JOB", NetworkWorkerThread::kDefaultMaxBlocksPerHddReadJob, 1);
```
What would happen if a `0` value is passed through the configuration option `MAX_BLOCKS_PER_HDD_READ_JOB`?

dmga44 replied:
The third parameter, which is 1, is the minimum value this option could get. So, if a 0 is passed, then the value used is 1, which current behavior in dev.

rolysr reviewed (approved):
LGTM. Take a look at my suggestions.

dmga44 committed:
test: Setup previous behavior in pread_eio tests

Signed-off-by: Dave <dave@leil.io>
